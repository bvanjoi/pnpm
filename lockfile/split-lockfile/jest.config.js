const config = require('../../jest.config.js')

module.exports = {
  ...config,
  testMatch: [...config.testMatch, "!**/test/utils.ts"]
}
