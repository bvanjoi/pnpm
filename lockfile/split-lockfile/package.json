{
  "name": "@pnpm/split-lockfile",
  "version": "0.1.0",
  "description": "Split and merge lockfile by importer",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "engines": {
    "node": ">=14.6"
  },
  "files": [
    "lib",
    "!*.map"
  ],
  "scripts": {
    "lint": "eslint src/**/*.ts test/**/*.ts",
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/lockfile/split-lockfile",
  "keywords": [
    "pnpm7",
    "pnpm",
    "shrinkwrap",
    "lockfile"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/lockfile/split-lockfile#readme",
  "devDependencies": {
    "@pnpm/split-lockfile": "workspace:*",
    "@types/ramda": "0.28.20",
    "ramda": "npm:@pnpm/ramda@0.28.1",
    "tempy": "^1.0.1"
  },
  "dependencies": {
    "@pnpm/lockfile-file": "workspace:*",
    "@pnpm/lockfile-types": "workspace:*",
    "@pnpm/types": "workspace:*",
    "@pnpm/lockfile-walker": "workspace:*",
    "@pnpm/merge-lockfile-changes": "workspace:*"
  },
  "funding": "https://opencollective.com/pnpm",
  "exports": {
    ".": "./lib/index.js"
  }
}
