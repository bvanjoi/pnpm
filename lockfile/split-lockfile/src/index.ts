export { splitLockfile } from './splitLockfile'
export { mergeSplittedLockfiles } from './mergeSplittedLockfiles'
