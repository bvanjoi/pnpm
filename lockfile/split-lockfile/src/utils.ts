import { Lockfile } from '@pnpm/lockfile-types'

type Info = Omit<Lockfile, 'importers' | 'lockfileVersion' | 'packages'>
function del (info: Info, property: keyof Info) {
  if (typeof info[property] === 'undefined') {
    delete info[property]
  }
}

export function pickLockfileInfo (lockfile: Lockfile): Info {
  const info = {
    neverBuiltDependencies: lockfile.neverBuiltDependencies,
    onlyBuiltDependencies: lockfile.onlyBuiltDependencies,
    overrides: lockfile.overrides,
    patchedDependencies: lockfile.patchedDependencies,
    time: lockfile.time,
    packageExtensionsChecksum: lockfile.packageExtensionsChecksum,
  }
  del(info, 'neverBuiltDependencies')
  del(info, 'onlyBuiltDependencies')
  del(info, 'overrides')
  del(info, 'patchedDependencies')
  del(info, 'time')
  del(info, 'packageExtensionsChecksum')
  return info
}
