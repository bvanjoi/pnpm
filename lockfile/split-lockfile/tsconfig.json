{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../__typings__/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../packages/types"
    },
    {
      "path": "../lockfile-file"
    },
    {
      "path": "../lockfile-types"
    },
    {
      "path": "../lockfile-walker"
    },
    {
      "path": "../merge-lockfile-changes"
    }
  ],
  "composite": true
}
