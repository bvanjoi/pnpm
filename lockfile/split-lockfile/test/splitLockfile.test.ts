import path from 'path'
import { splitLockfile } from '@pnpm/split-lockfile'
import { loadLockfile } from './utils'

test('splitLockfile() should works for single importer', async () => {
  const lockfile = await loadLockfile(path.resolve(__dirname, './fixture/single-import'))
  const result = splitLockfile(lockfile)
  expect(Object.keys(result).length).toBe(1)
  expect(result['.']).toStrictEqual(lockfile)
})

test('splitLockfile() should works for v5.4', async () => {
  const lockfile = await loadLockfile(path.resolve(__dirname, './fixture/v5.4'))
  const result = splitLockfile(lockfile)
  expect(Object.keys(result).length).toBe(4)
  expect(result['.']).toMatchSnapshot('packages/root')
  expect(result['packages/a']).toMatchSnapshot('packages/a')
  expect(result['packages/b']).toMatchSnapshot('packages/b')
  expect(result['packages/c']).toMatchSnapshot('packages/c')
})

test('splitLockfile() should works for v6', async () => {
  const lockfile = await loadLockfile(path.resolve(__dirname, './fixture/v6'))
  const result = splitLockfile(lockfile)
  expect(Object.keys(result).length).toBe(4)
  expect(result['.']).toMatchSnapshot('packages/root')
  expect(result['packages/a']).toMatchSnapshot('packages/a')
  expect(result['packages/b']).toMatchSnapshot('packages/b')
  expect(result['packages/c']).toMatchSnapshot('packages/c')
})

test('splitLockfile() should add info in root lockfile', async () => {
  const lockfile = await loadLockfile(path.resolve(__dirname, './fixture/info'))
  const result = splitLockfile(lockfile)
  expect(Object.keys(result).length).toBe(3)
  expect(result['.']).toMatchSnapshot('packages/root')
})
