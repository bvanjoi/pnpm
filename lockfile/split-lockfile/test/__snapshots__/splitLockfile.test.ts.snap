// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitLockfile() should add info in root lockfile: packages/root 1`] = `
{
  "importers": {
    ".": {
      "dependencies": undefined,
      "devDependencies": undefined,
      "optionalDependencies": undefined,
      "specifiers": {},
    },
  },
  "lockfileVersion": "6.0",
  "overrides": {
    "@babel/core": "7.20.0",
  },
}
`;

exports[`splitLockfile() should works for v5.4: packages/a 1`] = `
{
  "importers": {
    ".": {
      "dependencies": {
        "is-positive": "1.0.0",
      },
      "specifiers": {
        "is-positive": "1.0.0",
      },
    },
  },
  "lockfileVersion": 5.4,
  "packages": {
    "/is-positive/1.0.0": {
      "dev": false,
      "engines": {
        "node": ">=0.10.0",
      },
      "resolution": {
        "integrity": "sha512-xxzPGZ4P2uN6rROUa5N9Z7zTX6ERuE0hs6GUOc/cKBLF2NqKc16UwqHMt3tFg4CO6EBTE5UecUasg+3jZx3Ckg==",
      },
    },
  },
}
`;

exports[`splitLockfile() should works for v5.4: packages/b 1`] = `
{
  "importers": {
    ".": {
      "dependencies": {
        "@t/a": "link:../a",
        "is-positive": "1.0.0",
      },
      "devDependencies": {
        "is-negative": "1.0.0",
      },
      "specifiers": {
        "@t/a": "workspace:*",
        "is-negative": "1.0.0",
        "is-positive": "1.0.0",
      },
    },
  },
  "lockfileVersion": 5.4,
  "packages": {
    "/is-negative/1.0.0": {
      "dev": true,
      "engines": {
        "node": ">=0.10.0",
      },
      "resolution": {
        "integrity": "sha512-1aKMsFUc7vYQGzt//8zhkjRWPoYkajY/I5MJEvrc0pDoHXrW7n5ri8DYxhy3rR+Dk0QFl7GjHHsZU1sppQrWtw==",
      },
    },
  },
}
`;

exports[`splitLockfile() should works for v5.4: packages/c 1`] = `
{
  "importers": {
    ".": {
      "dependencies": {
        "is-negative": "1.0.1",
      },
      "specifiers": {
        "is-negative": "1.0.1",
      },
    },
  },
  "lockfileVersion": 5.4,
  "packages": {
    "/is-negative/1.0.1": {
      "dev": false,
      "engines": {
        "node": ">=0.10.0",
      },
      "resolution": {
        "integrity": "sha512-gmv+xZIqPGWfqTf195S+I8tw4h/VswI30Eu21e1r8qUVNHIiVdAVtpzoZBGQqrTNC+YjSwCNl33SXU7kuNwMzA==",
      },
    },
  },
}
`;

exports[`splitLockfile() should works for v5.4: packages/root 1`] = `
{
  "importers": {
    ".": {
      "specifiers": {},
    },
  },
  "lockfileVersion": 5.4,
}
`;

exports[`splitLockfile() should works for v6: packages/a 1`] = `
{
  "importers": {
    ".": {
      "dependencies": {
        "is-positive": "1.0.0",
      },
      "devDependencies": undefined,
      "optionalDependencies": undefined,
      "specifiers": {
        "is-positive": "1.0.0",
      },
    },
  },
  "lockfileVersion": "6.0",
  "packages": {
    "/is-positive/1.0.0": {
      "dev": false,
      "engines": {
        "node": ">=0.10.0",
      },
      "resolution": {
        "integrity": "sha512-xxzPGZ4P2uN6rROUa5N9Z7zTX6ERuE0hs6GUOc/cKBLF2NqKc16UwqHMt3tFg4CO6EBTE5UecUasg+3jZx3Ckg==",
      },
    },
  },
}
`;

exports[`splitLockfile() should works for v6: packages/b 1`] = `
{
  "importers": {
    ".": {
      "dependencies": {
        "@t/a": "link:../a",
        "is-positive": "1.0.0",
      },
      "devDependencies": {
        "is-negative": "1.0.0",
      },
      "optionalDependencies": undefined,
      "specifiers": {
        "@t/a": "workspace:*",
        "is-negative": "1.0.0",
        "is-positive": "1.0.0",
      },
    },
  },
  "lockfileVersion": "6.0",
  "packages": {
    "/is-negative/1.0.0": {
      "dev": true,
      "engines": {
        "node": ">=0.10.0",
      },
      "resolution": {
        "integrity": "sha512-1aKMsFUc7vYQGzt//8zhkjRWPoYkajY/I5MJEvrc0pDoHXrW7n5ri8DYxhy3rR+Dk0QFl7GjHHsZU1sppQrWtw==",
      },
    },
  },
}
`;

exports[`splitLockfile() should works for v6: packages/c 1`] = `
{
  "importers": {
    ".": {
      "dependencies": {
        "is-negative": "1.0.1",
      },
      "devDependencies": undefined,
      "optionalDependencies": undefined,
      "specifiers": {
        "is-negative": "1.0.1",
      },
    },
  },
  "lockfileVersion": "6.0",
  "packages": {
    "/is-negative/1.0.1": {
      "dev": false,
      "engines": {
        "node": ">=0.10.0",
      },
      "resolution": {
        "integrity": "sha512-gmv+xZIqPGWfqTf195S+I8tw4h/VswI30Eu21e1r8qUVNHIiVdAVtpzoZBGQqrTNC+YjSwCNl33SXU7kuNwMzA==",
      },
    },
  },
}
`;

exports[`splitLockfile() should works for v6: packages/root 1`] = `
{
  "importers": {
    ".": {
      "dependencies": undefined,
      "devDependencies": undefined,
      "optionalDependencies": undefined,
      "specifiers": {},
    },
  },
  "lockfileVersion": "6.0",
}
`;
